### ISSUE TITLE
Relations with multiple primary keys

### ISSUE THREAD

**Issue by andreasrueedlinger** [typeorm/typeorm#58]
I want to model a relation of two entities where the relation has some properties (eg. a Post has a many-to-many relation with Categories, but since the relation has some properties - like `order` or `categerizedAt` or whatever property you can imagine for such a relation - I have to introduce a join table I guess.
I have tried to get the correct mapping with typeorm but failed:
- how to define the primary key in the jointable?
- persisting fails with `[ERROR] [default] - Error: Entity PostCategoryRelation 
  has multiple primary keys. This operation is not supported on entities with 
  multiple primary keys                                                       
  at EntityMetadata.firstPrimaryColumn`

First try:

``` ts
@Table()
export class Post {

    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    title: string;

    @Column()
    text: string;

    @OneToMany(type => PostCategoryRelation, postCategoryRelation => postCategoryRelation.post)
    categories: PostCategoryRelation[];
}

@Table()
export class PostCategory {

    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    name: string;

    @OneToMany(type => PostCategoryRelation, postCategoryRelation => postCategoryRelation.category)
    posts: PostCategoryRelation[];
}

@Table()
export class PostCategoryRelation {
  @ManyToOne(type => Post, post => post.categories)
  @PrimaryColumn()
  post: Post;

  @ManyToOne(type => PostCategory, category => category.posts)
  @PrimaryColumn()
  category: PostCategory;

  @Column()
  addedByAdmin: boolean;

  @Column()
  addedByUser: boolean;
}
```

Did not work since PrimaryKey type cannot be determined.
Next try:

``` ts
@Table()
export class Post {

  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  title: string;

  @Column()
  text: string;

  @OneToMany(type => PostCategoryRelation, postCategoryRelation => postCategoryRelation.post)
  categories: PostCategoryRelation[];
}

@Table()
export class PostCategory {

  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  name: string;

  @OneToMany(type => PostCategoryRelation, postCategoryRelation => postCategoryRelation.category)
  posts: PostCategoryRelation[];
}

@Table()
export class PostCategoryRelation {

  @PrimaryColumn('int', { name: 'id_post' })
  private postRef: number;

  @PrimaryColumn('int', { name: 'id_cat' })
  private catRef: number;

  @ManyToOne(type => Post, post => post.categories)
  @JoinColumn({ name: 'id_post', referencedColumnName: 'id' })
  post: Post;

  @ManyToOne(type => PostCategory, category => category.posts)
  @JoinColumn({ name: 'id_cat', referencedColumnName: 'id' })
  category: PostCategory;

  @Column()
  addedByAdmin: boolean;

  @Column()
  addedByUser: boolean;
}
```

Now I get the above mentioned `[ERROR] [default] - Error: Entity PostCategoryRelation 
has multiple primary keys. This operation is not supported on entities with 
multiple primary keys                                                       
    at EntityMetadata.firstPrimaryColumn`

What is the right way to model this?

**Comment by pleerock** [comment]
quite complicated, I need to try try this sample by myself to figure out what is going on. Lets mark it as a bug for now, I'll return back to it later

**Comment by andreasrueedlinger** [comment]
Yes it might look a bit complicated. But for now I go with the solution to add an explicit primary key to all this "RelationEntities"

**Comment by 19majkel94** [comment]
I was talking about similar behavior in #43.
The solution is to create unique index on your foreign keys and create `@PrimaryGeneratedColumn()` in `PostCategoryRelation`

```typescript
@Table()
@Index((relation: PostCategoryRelation) => [relation.post, relation.category], { unique: true })
export class PostCategoryRelation {
  @PrimaryGeneratedColumn()
  public readonly id: number;

  @ManyToOne(type => Post, post => post.categories)
  post: Post;

  @ManyToOne(type => PostCategory, category => category.posts)
  category: PostCategory;

  @Column()
  addedByAdmin: boolean;

  @Column()
  addedByUser: boolean;
}
```
The simplest solution with `@ManyToOne(_type => Post, { primary: true })` doesn't work yet 👎

**Comment by pleerock** [comment]
fixed on my separate branch (you can try it if you want by `npm i typeorm@next`)

**Comment by 19majkel94** [comment]
Does this fix also fixes #43?

**Comment by pleerock** [comment]
@19majkel94 yes, it does
